### PR TITLE
Fix out of bounds read

### PR DESCRIPTION
**Pull request by Boneyan**
PR fixing bug:
```
panic: runtime error: index out of range [9] with length 0

goroutine 2182 [running]:
github.com/insomniacslk/dhcp/dhcpv4/nclient4.ipv4.protocol(...)
	/go/src/.../vendor/github.com/insomniacslk/dhcp/dhcpv4/nclient4/ipv4.go:108
github.com/insomniacslk/dhcp/dhcpv4/nclient4.ipv4.transportProtocol(...)
	/go/src/.../vendor/github.com/insomniacslk/dhcp/dhcpv4/nclient4/ipv4.go:124
github.com/insomniacslk/dhcp/dhcpv4/nclient4.(*BroadcastRawUDPConn).ReadFrom(0x400048a0f0, {0x40005f2600, 0x5dc, 0x1?})
	/go/src/.../vendor/github.com/insomniacslk/dhcp/dhcpv4/nclient4/conn_unix.go:115 +0x370
github.com/insomniacslk/dhcp/dhcpv4/nclient4.(*Client).receiveLoop(0x40000163f0)
	/go/src/.../vendor/github.com/insomniacslk/dhcp/dhcpv4/nclient4/client.go:262 +0x90
created by github.com/insomniacslk/dhcp/dhcpv4/nclient4.new
	/go/src/.../vendor/github.com/insomniacslk/dhcp/dhcpv4/nclient4/client.go:223 +0x3a4
```

### CODE REVIEW

**Comment by pmazzini** [comment]
Please sign the commit: https://github.com/insomniacslk/dhcp/pull/507/checks?check_run_id=15096874830

Example:
* In your local branch, run: git rebase HEAD~4 --signoff
* Force push your changes to overwrite the branch: git push --force-with-lease origin fix-out-of-bounds-read

**Comment by pmazzini** [comment]
Can you share some context on when you faced this out of bounds read?

**Comment by Boneyan** [comment]
App made dhcp requests in a mobile network and seems like received packet that caused panic.

File: dhcpv4/nclient4/conn_unix.go [hunk 1]
```diff
@@ -105,15 +105,18 @@ func (upc *BroadcastRawUDPConn) ReadFrom(b []byte) (int, net.Addr, error) {
...
 
-		if ipHdr.transportProtocol() != udpProtocolNumber {
-			continue
+		if headerLength > protocol {
+			if ipHdr.transportProtocol() != udpProtocolNumber {
+				continue
+			}
 		}
 
 		if !buf.Has(udpHdrLen) {
```
Comment (pmazzini) [inline dhcpv4/nclient4/conn_unix.go:122]: How about solving the issue by simply moving this if before the `udpProtocolNumber` check?

Comment (Boneyan) [reply dhcpv4/nclient4/conn_unix.go:122]: As I understand, `udpHdrLen` equals 8, so this won't solve the issue.

**Comment by pmazzini** [comment]
From https://github.com/insomniacslk/dhcp/blob/master/dhcpv4/nclient4/ipv4.go#L81

> ipv4 represents an ipv4 header stored in a byte array.
> Most of the methods of IPv4 access to the underlying slice without
> checking the boundaries and could panic because of 'index out of range'.
> Always call IsValid() to validate an instance of IPv4 before using other methods.

**Comment by pmazzini** [comment]
Maybe we should add the `IsValid` method:

https://github.com/google/gvisor/blob/master/pkg/tcpip/header/ipv4.go#L483

File: dhcpv4/nclient4/ipv4.go [hunk 2]
```diff
@@ -95,8 +97,40 @@ const (
...
+func ipVersion(b []byte) int {
+	// Length must be at least offset+length of version field.
+	if len(b) < versIHL+1 {
+		return -1
+	}
+	return int(b[versIHL] >> ipVersionShift)
+}
+
+// IsValid performs basic validation on the packet.
+func (b ipv4) isValid(pktSize int) bool {
```
Comment (pmazzini) [inline dhcpv4/nclient4/ipv4.go:116]: NIT: maybe move this below the encode for consistency with the other file you are copying it from

File: dhcpv4/nclient4/ipv4.go [hunk 3]
```diff
@@ -95,8 +97,40 @@ const (
...
+// if the packet is not large enough to contain the version field.
+func ipVersion(b []byte) int {
+	// Length must be at least offset+length of version field.
+	if len(b) < versIHL+1 {
+		return -1
+	}
+	return int(b[versIHL] >> ipVersionShift)
+}
+
+// IsValid performs basic validation on the packet.
```
Comment (pmazzini) [inline dhcpv4/nclient4/ipv4.go:115]: Lower case isValid

File: dhcpv4/nclient4/ipv4.go [hunk 4]
```diff
@@ -95,8 +97,40 @@ const (
 
 	// ipv4AddressSize is the size, in bytes, of an IPv4 address.
 	ipv4AddressSize = 4
+
+	// IPv4Version is the version of the IPv4 protocol.
+	ipv4Version = 4
 )
 
+// IPVersion returns the version of IP used in the given packet. It returns -1
```
Comment (pmazzini) [inline dhcpv4/nclient4/ipv4.go:105]: Lower case ipVerions

File: dhcpv4/nclient4/ipv4.go [hunk 5]
```diff
@@ -26,16 +27,17 @@ import (
...
+	tos            = 1
+	totalLen       = 2
+	id             = 4
+	flagsFO        = 6
+	ttl            = 8
+	protocol       = 9
+	checksumOff    = 10
+	srcAddr        = 12
+	dstAddr        = 16
+	ipVersionShift = 4
```
Comment (pmazzini) [inline dhcpv4/nclient4/ipv4.go:40]: Maybe move this to a separate group below `ipVersion` for consistency with the other file you are copying it from

**Comment by Boneyan** [comment]
Seems like current version of PR doesn't work (it can't receive response packet after offer)

**Comment by pmazzini** [comment]
> Seems like current version of PR doesn't work (it can't receive response packet after offer)

Do you know why?

**Comment by Boneyan** [comment]
I had such issue in old version of this PR when it was smth like
```
if headerLength < protocol {continue}  
//do udpheader check
```
and then I removed `continue` 
```
if headerLength > protocol {
    //do udpheader check
}
```
and it started working. So it has something to do with continue

**Comment by Boneyan** [comment]
I ran more tests and this version is actually working, previous test was misconducted

**Review by pmazzini — APPROVED** [review]

**Pull request merged**